akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/internal/ShardedDaemonProcessImpl.scala

```diff
@@ -0,0 +1,171 @@
+import akka.actor.typed.Behavior
+import akka.actor.typed.scaladsl.Behaviors
+import akka.actor.typed.scaladsl.LoggerOps
+import akka.annotation.InternalApi
+import akka.cluster.sharding.ShardRegion.EntityId
+import akka.cluster.sharding.typed.ClusterShardingSettings
+import akka.cluster.sharding.typed.ClusterShardingSettings.StateStoreModeDData
+import akka.cluster.sharding.typed.ShardingEnvelope
+import akka.cluster.sharding.typed.ShardingMessageExtractor
+import akka.cluster.sharding.typed.scaladsl
+import akka.cluster.sharding.typed.javadsl
+import akka.cluster.sharding.typed.ShardedDaemonProcessSettings
+import akka.cluster.sharding.typed.scaladsl.ClusterSharding
+import akka.cluster.sharding.typed.scaladsl.Entity
+import akka.cluster.sharding.typed.scaladsl.EntityTypeKey
+import akka.cluster.sharding.typed.scaladsl.StartEntity
+import akka.cluster.typed.Cluster
+import akka.japi.function
+import akka.util.PrettyDuration
+
+import scala.compat.java8.OptionConverters._
+import scala.concurrent.duration.Duration
+import scala.reflect.ClassTag
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] object ShardedDaemonProcessImpl {
+
+  object KeepAlivePinger {
+    sealed trait Event
+    case object Tick extends Event
+
+    def apply[T](
+        settings: ShardedDaemonProcessSettings,
+        name: String,
+        identities: Set[EntityId],
+        shardingRef: ActorRef[ShardingEnvelope[T]]): Behavior[Event] =
+      Behaviors.setup { context =>
+        Behaviors.withTimers { timers =>
+          def triggerStartAll(): Unit = {
+            identities.foreach(id => shardingRef ! StartEntity(id))
+          }
+
+          context.log.debugN(
+            s"Starting Sharded Daemon Process KeepAlivePinger for [{}], with ping interval [{}]",
+            name,
+            PrettyDuration.format(settings.keepAliveInterval))
+          timers.startTimerWithFixedDelay(Tick, settings.keepAliveInterval)
+          triggerStartAll()
+
+          Behaviors.receiveMessage {
+            case Tick =>
+              triggerStartAll()
+              context.log.debug("Periodic ping sent to [{}] processes", identities.size)
+              Behaviors.same
+          }
+        }
+      }
+  }
+
+  final class MessageExtractor[T] extends ShardingMessageExtractor[ShardingEnvelope[T], T] {
+    def entityId(message: ShardingEnvelope[T]): String = message match {
+      case ShardingEnvelope(id, _) => id
+    }
+
+    def shardId(entityId: String): String = entityId
+
+    def unwrapMessage(message: ShardingEnvelope[T]): T = message.message
+  }
+
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] final class ShardedDaemonProcessImpl(system: ActorSystem[_])
+    extends javadsl.ShardedDaemonProcess
+    with scaladsl.ShardedDaemonProcess {
+
+  import ShardedDaemonProcessImpl._
+
+  def init[T](name: String, numberOfInstances: Int, behaviorFactory: Int => Behavior[T])(
+      implicit classTag: ClassTag[T]): Unit = {
```

Comment:
Should it return the region `ActorRef[ShardingEnvelope[T]`, which can be used for sending other messages to the actors?

Reply:
The caller doesn't know how we name them so that is one reason to not do that.

Reply:
Did we agree on what to do here? I think for now let's just return Unit and users can use receptionist. We can always return ref to send messages to alter

Reply:
I'd prefer keeping it a `Unit`.

Reply:
ok, unit it is